<a href="https://colab.research.google.com/github/SergioAmado1408/Taller--2/blob/Tallerr/2_Taller.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Taller 2

Ana Garzón,
Sergio Amado

Librerias

In [30]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.optimize import minimize
from scipy.optimize import linprog  
import cvxpy as cp


Punto 1

In [ ]:
Distribucion = pd.DataFrame({' ' : ['Insumos','Mano de obra', 'Capital', 'Ganancia x u'],
        'Producto A' : [1, 2, 1, 3],'Producto B' : [1, 4, 2, 5]})
Distribucion

Producto A  Producto B
0       Insumos           1           1
1  Mano de obra           2           4
2       Capital           1           2
3  Ganancia x u           3           5

In [ ]:
from pandas.core.arrays.sparse.dtype import astype_nansafe
c = np.array([3, 5])

A = np.array([[1, 1],
                  [2,4],
                    [1,2]])
B = np.array([5000,1000,2000])

bandas = [(100, None),
           (150, None)]

# Solve the problem:
ans = linprog(-c, A_ub=A, b_ub=B,
                  bounds=bandas, method='revised simplex')

ans

<ipython-input-5-d660959e5903>:13: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  ans = linprog(-c, A_ub=A, b_ub=B,


 message: Optimization terminated successfully.
 success: True
  status: 0
     fun: -1350.0
       x: [ 2.000e+02  1.500e+02]
     nit: 2

Para obtener una utilidad de $1350, se necesitan 150 unidades de B y 200 deA

Puunto 2




In [ ]:
funcion=np.array([0.10,0.15,0.16,0.08])

A= np.array([[0.5,0.3,0.25,0.6],
                [0.3,0.1,0.4,0.2],
                [0.2,0.6,0.35,0.2]])

B=np.array([0.35,0.30,0.15])

C=np.array([(1,1,1,1)])

D=np.array([1])

ans=linprog(-funcion,A_ub=-A,b_ub=-B,A_eq=C,b_eq=D,method="simplex")
pd.DataFrame(ans.x).round(3).T

<ipython-input-6-79414192fa52>:13: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  ans=linprog(-funcion,A_ub=-A,b_ub=-B,A_eq=C,b_eq=D,method="simplex")


0      1      2      3
0  0.0  0.158  0.579  0.263

Por CVXPY

In [ ]:
n=4

x = cp.Variable(n)

objetivo = cp.Minimize( -funcion.T@x)

restriccion=[cp.sum(x)==1,A@x >=B,x>=0]

problema=cp.Problem(objetivo,restriccion)
result=problema.solve()

seccion2=pd.DataFrame(x.value)
pd.DataFrame(seccion2).round(3).T

print("la mayor utilidad es de", problema.value)
print(round(pd.DataFrame(x.value),3))

la mayor utilidad es de -0.13736842087092851
       0
0  0.000
1  0.158
2  0.579
3  0.263


La utilidad maxxima es de 13,73 y para obtenerla se debe invertir 15.7% en el fondo 1, 57,8% en el fondo 2 y 26.3% en el fondo 3.

  Punto 3

In [ ]:
# Parameters 
r = 0.01
rr = 0.02
rrr = 0.003

# Restricciones de desigualdad
c= np.array([0,0,0,0,0,0,0,0,0,0,0,0,0,1])

A= np.array([[1,0,0,0,0,1,0,0,-1,0,0,0,0,0],
              [-1-r,1,0,0,0,0,1,0,1+rrr,-1,0,0,0,0],
              [0,-1-r,1,0,0,0,0,1,0,1+rrr,-1,0,0,0],
              [0,0,-1-r,1,0,-1-rr,0,0,0,0,1+rrr,-1,0,0],
              [0,0,0,-1-r,1,0,-1-rr,0,0,0,0,1+rrr,-1,0],
              [0,0,0,0,-1-r,0,0,-1-rr,0,0,0,0,1+rrr,-1]])

B=np.array([150,100,-200,200,-50,-300])

C = np.array([[1,1,1,1,1,1,1,1,1,1,1,1,1,1]])

D = np.array([100])

ans = linprog(-c, A_ub=-C, b_ub=-D, A_eq=A, b_eq=B, method='revised simplex') 

pd.DataFrame(ans.x).round(3)

<ipython-input-15-c54cfc5d4308>:22: DeprecationWarning: `method='revised simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  ans = linprog(-c, A_ub=-C, b_ub=-D, A_eq=A, b_eq=B, method='revised simplex')


0
0     0.000
1     0.000
2     0.000
3     0.000
4    52.000
5   150.000
6   100.000
7   151.944
8     0.000
9     0.000
10  351.944
11    0.000
12    0.000
13   92.497

Por CVXPY

In [ ]:
n=14
x = cp.Variable(n)
funcion = cp.Minimize(-c3.T@x)
restriccion = [x>=0,A@x>=B]

problema=cp.Problem(funcion,restriccion)
resultado=problema.solve()

print("Maxima Utilidad es de", -problema.value)

print(round(pd.DataFrame(x.value),4))

Maxima Utilidad es de 92.49694912403417
           0
0     0.0000
1    41.0006
2     0.0000
3     0.0000
4    10.1794
5   150.0000
6    58.9994
7   193.3547
8     0.0000
9     0.0000
10  351.9442
11    0.0000
12    0.0000
13   92.4969


El mayor saldo es de 92.4969 millone en el mes 6

Punto 4

In [ ]:
Funcion =np.array([0, 0, 0, 0, 0, 0, 0, 1])

c = (1+0.0001)**7

A = np.array([[1, 0, 0, 0, -1, 0, 0, 0],
              [0, 1, 0, 0, c, -1, 0, 0],
              [0, 0, 1, 0, 0, c, -1, 0],
              [0, 0, 0, 1, 0, 0, c, -1]])

B = np.array([[150, -100, -50, 200]])

ans= linprog(-Funcion, A_ub=A, b_ub=-B, method='simplex')
pd.DataFrame(ans.x).round(3).T



<ipython-input-23-59cc6e17a720>:12: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  ans= linprog(-Funcion, A_ub=A, b_ub=-B, method='simplex')


0    1    2    3      4       5     6       7
0  0.0  0.0  0.0  0.0  150.0  50.105  0.14  200.14

In [ ]:
ans

 message: Optimization failed. The problem appears to be unbounded.
 success: False
  status: 3
     fun: -200.14021366142958
       x: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00  1.500e+02
            5.011e+01  1.401e-01  2.001e+02]
     nit: 4

La utilidad maxima es de 0.15 millones de pesos.

Punto 5

In [ ]:
c = np.array([0,0,0,5,8,10])

A = np.array([[0,0,0,1,0,0],[0,1,0,0,0,0],[0,0,1,0,0,0],[0,0,0,0,1,0],[0,0,0,0,0,1]])

B=np.array([2000,300,250,2000,6000])

C=np.array([[1,1,1,0,0,0], [0,0,0,1,1,1]])
D= np.array([1000,10000])

ans= linprog(-c, A_ub=A, b_ub=B, A_eq=C, b_eq=D, method="simplex")

print(-ans.fun)
pd.DataFrame(ans.x).round(3).T

86000.0


<ipython-input-26-bc8d11854533>:10: DeprecationWarning: `method='simplex'` is deprecated and will be removed in SciPy 1.11.0. Please use one of the HiGHS solvers (e.g. `method='highs'`) in new code.
  ans= linprog(-c, A_ub=A, b_ub=B, A_eq=C, b_eq=D, method="simplex")


0      1      2       3       4       5
0  450.0  300.0  250.0  2000.0  2000.0  6000.0

La mayor utilidad posible es de $86000, invirtiendo 2000 en ventas, 2000 en marketing y 
6000 en producción.

Punto 6

In [44]:
c = np.array([[4,5,6,8,10],[6,4,3,5,8],[9,7,4,3,4]])
b = np.array([500, 500, 500])

d = np.array ( [80, 270, 250, 160, 180])
x=cp.Variable((3,5),nonneg=True)
m=np.array([1000, 1000, 1000])
h= cp.Variable(3,boolean=True)
obje = cp.sum(cp.multiply (c, x)) 

objetivo = cp.Minimize(obje)
restricciones=[cp.sum (x, axis=1)<=b@y,cp.sum(x,axis=0)==d,cp.sum(h*b)>=cp.sum(d)]
cp.Problem(objetivo, restricciones)

problema=cp.Problem(objetivo,restricciones)
problema.solve()
print (round(pd.DataFrame(x.value),2))



      0      1      2      3      4
0  80.0   -0.0   -0.0   -0.0   -0.0
1  -0.0  270.0  250.0   -0.0   -0.0
2  -0.0   -0.0   -0.0  160.0  180.0


/usr/local/lib/python3.9/dist-packages/cvxpy/expressions/expression.py:593: UserWarning: 
This use of ``*`` has resulted in matrix multiplication.
Using ``*`` for matrix multiplication has been deprecated since CVXPY 1.1.
    Use ``*`` for matrix-scalar and vector-scalar multiplication.
    Use ``@`` for matrix-matrix and matrix-vector multiplication.
    Use ``multiply`` for elementwise multiplication.
This code path has been hit 11 times so far.

  warnings.warn(msg, UserWarning)
